crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -164,11 +167,41 @@ where
                 aux_store,
                 next_key_index: AtomicU16::new(next_key_index),
                 cache: Mutex::new(cache),
+                initialization_completed: Mutex::new(false),
             }),
             confirmation_depth_k,
         })
     }
 
+    /// Waits until the initialization completed with timeout.
+    pub async fn wait_for_initialization(&self, timeout: Duration) -> bool {
+        let start = Instant::now();
+        const WAIT_PERIOD: Duration = Duration::from_secs(1);
+
+        loop {
+            if *self.inner.initialization_completed.lock() {
+                return true;
+            }
+
+            if Instant::now().duration_since(start) > timeout {
+                return false;
+            }
+
+            debug!(
+                ?start,
+                ?timeout,
+                "Waiting for the segment headers store initialization..."
+            );
+            tokio::time::sleep(WAIT_PERIOD).await;
+        }
+    }
```

Comment:
No `sleep`, please. Code should be reactive and I have shown how to do it during our sync call.

This is a very awkward API with two public methods that you are supposed to call in a specific order at specific time, it is basically asking for trouble to happen.

Reply:
I don't think I remember the synchronization part of our discussion. AFAIR we were mostly focused on other aspects like "making future faster", "archiver autoinitialization", etc.
Regarding "sleep" and "2 API methods": I initially thought about either protecting (by lock) every method of the segment headers store or using the special primitive (like tokio notify) but chose to synchronize only places where we need it. Do you have in mind 2-3 lines of pseudocode to describe your preferred solution?

Reply:
What I proposed above will only initialize archiver once in node lifetime for genesis segment, which is fine. All other restarts will not result in extra delays in node initialization, which is exactly what we want. Removing that single delay with awkward API is not worth it IMO.